### PR TITLE
Index/Term's in raft state should be locked/get/set as a single thing

### PR DESCRIPTION
**Pull request by superfell**
fixes Issue #84 item 1

### CODE REVIEW

File: raft.go [hunk 1]
```diff
@@ -1821,7 +1818,8 @@ func (r *Raft) compactLogs(snapIdx uint64) error {
 	}
 
 	// Check if we have enough logs to truncate
-	if r.getLastLogIndex() <= r.conf.TrailingLogs {
+	lastLogIdx, _ := r.getLastLog()
```
Comment (ongardie) [inline raft.go]: Nit: I think this could be r.getLastLogIndexOnly()

File: raft.go [hunk 2]
```diff
@@ -1367,7 +1367,7 @@ func (r *Raft) appendEntries(rpc RPC, a *AppendEntriesRequest) {
 		last := a.Entries[n-1]
 
 		// Delete any conflicting entries
-		lastLogIdx := r.getLastLogIndex()
+		lastLogIdx, _ := r.getLastLog()
```
Comment (ongardie) [inline raft.go]: Nit: getLastLogIndexOnly() here too?

**Comment by ongardie** [comment]
This PR looks to me like a step in the right direction. I'm still worried about shared state across the library, but this PR will help with some issues. +1

File: state.go [hunk 3]
```diff
@@ -83,20 +87,39 @@ func (r *raftState) setCurrentTerm(term uint64) {
...
-	return atomic.LoadUint64(&r.LastLogIndex)
+func (r *raftState) getLastLog() (index, term uint64) {
+	r.lastLock.Lock()
+	index = r.lastLogIndex
+	term = r.lastLogTerm
+	r.lastLock.Unlock()
+	return
+}
+
+func (r *raftState) getLastLogIndexOnly() uint64 {
```
Comment (slackpad) [inline state.go]: This one is a little funky - I'd prefer just having all callers call `log, _ := getLastLog()`.

File: raft.go [hunk 4]
```diff
@@ -1829,7 +1827,7 @@ func (r *Raft) compactLogs(snapIdx uint64) error {
 	// back from the head, which ever is further back. This ensures
 	// at least `TrailingLogs` entries, but does not allow logs
 	// after the snapshot to be removed.
-	maxLog := min(snapIdx, r.getLastLogIndex()-r.conf.TrailingLogs)
+	maxLog := min(snapIdx, lastLogIdx-r.conf.TrailingLogs)
```
Comment (slackpad) [inline raft.go]: This was a good catch.

**Comment by slackpad** [comment]
Just one tiny comment on this one. We should target this to https://github.com/hashicorp/raft/tree/issue-84-integration, and a rebase should make Travis CI less flaky with ancient Go 1.2 removed.

File: state.go (hunk 3 above)
Comment (slackpad) [inline state.go]: Looking through the history now I realize you had it this way originally :-)

@ongardie were you worried about being able to more easily find places that don't fetch both together or was this just a style thing? It's not a huge deal either way but `idx, _` seems more idiomatic.

Comment (superfell) [inline state.go]: Yeah,i started with just the 2 item return version, but adding index only simplified its usage in replication.go. am happy to go back to just the 2 item version if you'd rather have just that.

Comment (slackpad) [inline state.go]: I'd prefer only the 2 item version. I think for future maintenance it's probably better to make people think about fetching them together as a unit as well.

**Comment by ongardie** [comment]
I think this one should go into master. It's pretty harmless, so there's no need to stage it in the issue-84-integration branch only.

**Comment by ongardie** [comment]
Heads up: this'll probably conflict with #101 but should be easy to resolve.

**Comment by armon** [comment]
Reviewed. I agree with the existing comments, then LGTM!

**Comment by superfell** [comment]
rebased to current master, lastLogIndexOnly() removed.

**Comment by armon** [comment]
LGTM!

**Pull request merged**